main/src/com/google/refine/grel/controls/ForRange.java

```diff
@@ -86,27 +83,31 @@ public Object call(Properties bindings, Evaluable[] args) {
                 long step = ((Number) stepO).longValue();
                 double to = ((Number) toO).doubleValue();
 
-                while (from < to) {
-                    bindings.put(indexName, from);
+                if (!((from < to && step > 0) || (from > to && step < 0))) {
+                    return new EvalError(EvalErrorMessage.invalid_arg());
+                }
 
+                int rangeSize = (int) Math.ceil(Math.abs(from - to) / Math.abs(step));
+                for (int i = 0; i < rangeSize; i++) {
+                    bindings.put(indexName, from);
                     Object r = args[4].evaluate(bindings);
```

Comment:
I think the way that this was originally organized without the intermediate `rangeSize` calculation is more direct and avoids any additional loss of precision.